src/main/java/net/devtech/arrp/json/loot/JFunction.java

```diff
@@ -1,28 +1,60 @@
 package net.devtech.arrp.json.loot;
 
+import com.google.gson.*;
+import net.minecraft.util.Identifier;
+
 import java.lang.reflect.Type;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
-
-import com.google.gson.JsonElement;
-import com.google.gson.JsonObject;
-import com.google.gson.JsonSerializationContext;
-import com.google.gson.JsonSerializer;
 
 public class JFunction implements Cloneable {
 	private final List<JCondition> conditions = new ArrayList<>();
-	private Map<String, Object> properties = new HashMap<>();
+	private JsonObject properties = new JsonObject();
 
 	/**
 	 * @see JLootTable#function(String)
 	 */
 	public JFunction(String function) {
-		this.properties.put("function", function);
+		function(function);
+	}
+
+	public JFunction function(String function) {
+		this.properties.addProperty("function", function);
+		return this;
+	}
+
+	public JFunction set(JsonObject properties) {
+		properties.addProperty("function",this.properties.get("function").getAsString());
+		this.properties = properties;
+		return this;
+	}
+
+	public JFunction parameter(String key, JsonElement value) {
+		this.properties.add(key, value);
+		return this;
+	}
+
+	public JFunction parameter(String key, String value) {
+		return parameter(key, new JsonPrimitive(value));
+	}
+
+	public JFunction parameter(String key, Number value) {
+		return parameter(key, new JsonPrimitive(value));
+	}
+
+	public JFunction parameter(String key, Boolean value) {
+		return parameter(key, new JsonPrimitive(value));
+	}
+
+	public JFunction parameter(String key, Identifier value) {
+		return parameter(key, value.toString());
+	}
+
+	public JFunction parameter(String key, Character value) {
+		return parameter(key, new JsonPrimitive(value));
 	}
 
-	public JFunction add(JCondition condition) {
+	public JFunction condition(JCondition condition) {
```

Comment:
breaks backwards compatibility, deprecate `add` the old method and make it delegate to condition

Reply:
fixed